### PR TITLE
Adopt `ruff` as the central tool for linting, formatting, and import sorting

### LINKED ISSUE #592: [DevOps]: Consider replacing `flake8` linter, `black` formatter, and `isort` with `ruff`
I saw `xarray` uses `ruff` for linting. It is supposedly much faster than `flake8`.

- https://github.com/astral-sh/ruff
- https://pythonspeed.com/articles/pylint-flake8-ruff/
- https://www.reddit.com/r/Python/comments/12ha6mc/ruff_one_python_linter_to_rule_them_all/

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Description

- Closes #592 

### [Why ruff?](https://docs.astral.sh/ruff/)
- Meets most of our needs out-of-the-box (linter, formatter, import sorting)
- Adopting ruff simplifies DevOps tooling and configuration
- It is **significantly** faster for linting, formatting, and import sorting compared to `flake8`, `black`, and `isort`
- Mostly interoperable and configured to replicate  `flake8`, `black`, and `isort` out-of-the-box
- Is easily extensible with tools like `flake8-docstrings` and `flake8-bugbear` 
- Rapidly growing popularity on [GitHub](https://github.com/astral-sh/ruff), adopted by other [major Python packages](https://github.com/astral-sh/ruff?tab=readme-ov-file#whos-using-ruff)

### Summary of changes
- Replaces `flake8`, `black`, and `isort`/`flake8-isort` with a single tool, `ruff` in conda dev environment and `pre-commit`
- Update contributing docs to refer to `ruff`
- Fix linting issues including:
  - [B007 - unused-loop-control-variable](https://docs.astral.sh/ruff/rules/unused-loop-control-variable/) 
  - [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)
  - [B006 - mutable-argument default](https://docs.astral.sh/ruff/rules/mutable-argument-default/)
    - I just learned about this one. We default various API arguments to a list (e.g., `axes=["X", "Y"]`). It is bad practice to set default arguments to a mutable object: https://docs.astral.sh/ruff/rules/mutable-argument-default/#why-is-this-bad
    - Instead, non-mutable objects such as `tuples` should be used
    - https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument
- Ignored linting issues including:
  - [B018 - useless-expression](https://docs.astral.sh/ruff/rules/useless-expression/) not ignored in tests that expect exceptions
  - [F401 - unused-import](https://docs.astral.sh/ruff/rules/unused-import/) not ignored in `regridder/__init__.py`
- Run formatter

### What do contributors need to do?
1. Recreate their development environment after this PR is merged
   - `conda env create -f conda/dev.yml`
2. Reinstall pre-commit hooks into their `xcdat` repo: `pre-commit install`
   - `pre-commit install`

### Future TODOs
- Enable `flake8-docstring` for `ruff` in `pyproject.toml` and fix docstring issues

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: tests/test_bounds.py [hunk 1]
```diff
@@ -847,7 +847,7 @@ def test_add_monthly_bounds_for_end_of_month_set_to_true(self):
         # Get time axis and create new axis with time set to first day of month.
         time = ds_with_bnds.time
         new_time = []
-        for i, t in enumerate(time.values):
+        for _, t in enumerate(time.values):
```
Comment (tomvothecoder) [inline tests/test_bounds.py:850]: Fixes [B007 - unused-loop-control-variable](https://docs.astral.sh/ruff/rules/unused-loop-control-variable/)

File: tests/test_regrid.py [hunk 2]
```diff
@@ -663,7 +668,7 @@ def test_map_latitude_coarse_to_fine(self):
             np.testing.assert_allclose(x, y)
 
         for x2, y2 in zip(weights, expected_weigths):
-            np.testing.assert_allclose(x, y)
+            np.testing.assert_allclose(x2, y2)
```
Comment (tomvothecoder) [inline tests/test_regrid.py:671]: Fixes [B007 - unused-loop-control-variable](https://docs.astral.sh/ruff/rules/unused-loop-control-variable/)

File: tests/test_regrid.py [hunk 3]
```diff
@@ -1254,7 +1259,7 @@ def test_grid(self):
             ValueError,
             match=r".*lon\d?.*lon\d?.*",
         ):
-            ds_multi.regridder.grid
+            ds_multi.regridder.grid  # noqa: B018
```
Comment (tomvothecoder) [inline tests/test_regrid.py:1262]: Ignores [B018 - useless-expression](https://docs.astral.sh/ruff/rules/useless-expression/)

File: xcdat/dataset.py [hunk 4]
```diff
@@ -44,7 +44,7 @@
 def open_dataset(
     path: str | os.PathLike[Any] | BufferedIOBase | AbstractDataStore,
     data_var: Optional[str] = None,
-    add_bounds: List[CFAxisKey] | None = ["X", "Y"],
+    add_bounds: List[CFAxisKey] | Tuple[CFAxisKey, ...] | None = ("X", "Y"),
```
Comment (tomvothecoder) [inline xcdat/dataset.py:47]: Fixes [B006 - mutable-argument default](https://docs.astral.sh/ruff/rules/mutable-argument-default/)

File: xcdat/bounds.py [hunk 5]
```diff
@@ -124,7 +125,7 @@ def keys(self) -> List[str]:
         )
 
     def add_missing_bounds(  # noqa: C901
-        self, axes: List[CFAxisKey] = ["X", "Y", "T"]
+        self, axes: List[CFAxisKey] | Tuple[CFAxisKey, ...] = ("X", "Y", "T")
     ) -> xr.Dataset:
```
Comment (tomvothecoder) [inline xcdat/bounds.py:131]: Fixes [B006 - mutable-argument default](https://docs.astral.sh/ruff/rules/mutable-argument-default/)

File: xcdat/dataset.py [hunk 6]
```diff
@@ -127,7 +127,7 @@ def open_dataset(
 def open_mfdataset(
     paths: str | NestedSequence[str | os.PathLike],
     data_var: Optional[str] = None,
-    add_bounds: List[CFAxisKey] | None = ["X", "Y"],
+    add_bounds: List[CFAxisKey] | Tuple[CFAxisKey, ...] | None = ("X", "Y"),
```
Comment (tomvothecoder) [inline xcdat/dataset.py:130]: Fixes [B006 - mutable-argument default](https://docs.astral.sh/ruff/rules/mutable-argument-default/)

File: xcdat/spatial.py [hunk 7]
```diff
@@ -68,7 +69,7 @@ def __init__(self, dataset: xr.Dataset):
     def average(
         self,
         data_var: str,
-        axis: List[SpatialAxis] = ["X", "Y"],
+        axis: List[SpatialAxis] | Tuple[SpatialAxis, ...] = ("X", "Y"),
```
Comment (tomvothecoder) [inline xcdat/spatial.py:74]: Fixes [B006 - mutable-argument default](https://docs.astral.sh/ruff/rules/mutable-argument-default/)

File: xcdat/regridder/accessor.py [hunk 8]
```diff
@@ -198,7 +198,7 @@ def horizontal(
         except KeyError as e:
             raise ValueError(
                 f"Tool {e!s} does not exist, valid choices {list(HORIZONTAL_REGRID_TOOLS)}"
-            )
+            ) from e
```
Comment (tomvothecoder) [inline xcdat/regridder/accessor.py:201]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/regridder/accessor.py [hunk 9]
```diff
@@ -269,7 +269,7 @@ def vertical(
             raise ValueError(
                 f"Tool {e!s} does not exist, valid choices "
                 f"{list(VERTICAL_REGRID_TOOLS)}"
-            )
+            ) from e
```
Comment (tomvothecoder) [inline xcdat/regridder/accessor.py:272]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/regridder/regrid2.py [hunk 10]
```diff
@@ -555,8 +555,8 @@ def _get_dimension(input_data_var, cf_axis_name):
 def _get_bounds_ensure_dtype(ds, axis):
     try:
         name = ds.cf.bounds[axis][0]
-    except (KeyError, IndexError):
-        raise RuntimeError(f"Could not determine {axis!r} bounds")
+    except (KeyError, IndexError) as e:
+        raise RuntimeError(f"Could not determine {axis!r} bounds") from e
```
Comment (tomvothecoder) [inline xcdat/regridder/regrid2.py:559]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/regridder/xgcm.py [hunk 11]
```diff
@@ -148,8 +148,10 @@ def vertical(self, data_var: str, ds: xr.Dataset) -> xr.Dataset:
         """See documentation in :py:func:`xcdat.regridder.xgcm.XGCMRegridder`"""
         try:
             output_coord_z = get_dim_coords(self._output_grid, "Z")
-        except KeyError:
-            raise RuntimeError("Could not determine 'Z' coordinate in output dataset")
+        except KeyError as e:
+            raise RuntimeError(
+                "Could not determine 'Z' coordinate in output dataset"
+            ) from e
```
Comment (tomvothecoder) [inline xcdat/regridder/xgcm.py:154]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/regridder/xgcm.py [hunk 12]
```diff
@@ -165,11 +167,11 @@ def vertical(self, data_var: str, ds: xr.Dataset) -> xr.Dataset:
             target_data = ds[self._target_data]
         except ValueError:
             target_data = self._target_data
-        except KeyError:
+        except KeyError as e:
             if self._target_data is not None and isinstance(self._target_data, str):
                 raise RuntimeError(
                     f"Could not find target variable {self._target_data!r} in dataset"
-                )
+                ) from e
```
Comment (tomvothecoder) [inline xcdat/regridder/xgcm.py:174]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/regridder/xgcm.py [hunk 13]
```diff
@@ -216,8 +218,8 @@ def _get_grid_positions(self) -> Dict[str, Union[Any, Hashable]]:
 
         try:
             coord_z = get_dim_coords(self._input_grid, "Z")
-        except KeyError:
-            raise RuntimeError("Could not determine `Z` coordinate in dataset.")
+        except KeyError as e:
+            raise RuntimeError("Could not determine `Z` coordinate in dataset.") from e
```
Comment (tomvothecoder) [inline xcdat/regridder/xgcm.py:222]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/regridder/xgcm.py [hunk 14]
```diff
@@ -231,8 +233,8 @@ def _get_grid_positions(self) -> Dict[str, Union[Any, Hashable]]:
 
         try:
             bounds_z = self._input_grid.bounds.get_bounds("Z")
-        except KeyError:
-            raise RuntimeError("Could not determine `Z` bounds in dataset.")
+        except KeyError as e:
+            raise RuntimeError("Could not determine `Z` bounds in dataset.") from e
```
Comment (tomvothecoder) [inline xcdat/regridder/xgcm.py:237]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

File: xcdat/temporal.py [hunk 15]
```diff
@@ -959,12 +960,12 @@ def _is_valid_reference_period(self, reference_period: Tuple[str, str]):
...
             datetime.strptime(reference_period[0], "%Y-%m-%d")
             datetime.strptime(reference_period[1], "%Y-%m-%d")
-        except (IndexError, ValueError):
+        except (IndexError, ValueError) as e:
             raise ValueError(
                 "Reference periods must be a tuple of strings with the format "
                 "'yyyy-mm-dd'. For example, reference_period=('1850-01-01', "
                 "'1899-12-31')."
-            )
+            ) from e
```
Comment (tomvothecoder) [inline xcdat/temporal.py:968]: Fixes [B904 - raise-without-from-inside-except](https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/#raise-without-from-inside-except-b904)

**Review by tomvothecoder — COMMENTED** [review]
My self-review

**Comment by tomvothecoder** [comment]
@xCDAT/core-developers happy the get thoughts if any of you care about this like I do.

All you need to do on your end is reinstall your conda dev env and `pre-commit` hooks if this PR is merged

**Comment by pochedls** [comment]
I trust you.

**Pull request merged**